bop_toolkit_lib/renderer_batch.py

```diff
@@ -0,0 +1,161 @@
+                setattr(args, key, list(value))
+            if key == "vsd_normalized_by_diameter":
+                setattr(args, key, bool(value))
+            if key == "step":
+                setattr(args, key, str(value))
+            if key == "obj_id":
+                setattr(args, key, int(value))
+            else:
+                setattr(args, key, value)
+        return args
+
+
+class BatchRenderer:
+    """
+    Batch renderer for renderer.create_renderer
+    """
+
+    def __init__(
+        self,
+        width,
+        height,
+        renderer_type="cpp",
+        mode="rgb+depth",
+        shading="phong",
+        bg_color=(0.0, 0.0, 0.0, 0.0),
+        num_workers=1,
+        tmp_dir="/tmp/vsd/",
+    ):
+        assert num_workers >= 1
+        self.num_workers = num_workers
+        self.tmp_dir = tmp_dir
+        self.rendere_args = dict(
+            width=width,
+            height=height,
+            render_type=renderer_type,
+            mode=mode,
+            shading=shading,
+            bg_color=bg_color,
+        )
+        self.models = {}
+
+    def add_object(self, obj_id, model_path):
+        self.models[obj_id] = model_path
+
+    def _init_renderers(self):
+        # create folder for each worker
+        for worker_id in range(self.num_workers):
+            worker_folder = os.path.join(self.tmp_dir, f"vsd_worker_{worker_id}")
+            os.makedirs(worker_folder, exist_ok=True)
+
+        # save renderer args
+        renderer_args_path = os.path.join(self.tmp_dir, "renderer_args.json")
+        inout.save_json(renderer_args_path, self.rendere_args)
+
+        # save model paths
+        models_dict_path = os.path.join(self.tmp_dir, "models_dict.json")
+        inout.save_json(models_dict_path, self.models)
+
+    def run_vsd(self, all_im_errs):
+        """
+        Compute VSD errors for all POSE_ERROR_VSD_ARGS
+        """
+        # call init_renderers for all workers, after adding all objects
+        self._init_renderers()
+
+        in_counter = 0
+        for idx_im, im_errs in enumerate(all_im_errs):
+            for idx_err, err in enumerate(im_errs):
+                for gt_id in err["errors"].keys():
+                    if isinstance(err["errors"][gt_id], POSE_ERROR_VSD_ARGS):
+                        worker_id = in_counter % self.num_workers
+                        vsd_args_path = os.path.join(
+                            self.tmp_dir,
+                            f"vsd_worker_{worker_id}/{idx_im}_{idx_err}_{gt_id}.npz",
+                        )
+                        err["errors"][gt_id].to_file(vsd_args_path)
+                        in_counter += 1
+
+        all_im_errs = self.start_run_vsd(all_im_errs, in_counter)
+        return all_im_errs
+
+    def start_run_vsd(self, all_im_errs, in_counter):
+        cmds = []
+        for worker_id in range(self.num_workers):
+            cmd = [
+                "python",
+                "bop_toolkit_lib/call_renderer.py",
```

Comment:
Is parallel calling of this `call_renderer` script (which always needs to initiate the rendering buffer, load object model, etc. from scratch) really faster than just running a single renderer (that can be initialized only once at the beginning; as before this diff) in sequence? If not or if the difference is not significant (which I assume) I would stick to calculating VSD in sequence (reason: the code is simpler to understand and maintain).

Reply:
As shown in the table above, the run-time difference is significant for BOP5 datasets. This is because (1) the initialisation is not very long (<10s) while all the workers are run in parallel, and (2) the number of images to render is much larger than the number of objects to initialize + the renderer buffer . However, I got your points so I add [this check](https://github.com/nv-nguyen/bop_toolkit/blob/master/bop_toolkit_lib/renderer_batch.py#L95) which will make sure that the multiprocessing is done when it is need.

Reply:
To make the code more readable and modular, it would be great if we could clearly divide the renderer and the error computation code. For example, 
- defining POSE_ERROR_VSD_ARGS in `bop_toolkit_lib/pose_error.py`
- rename `call_renderer.py` to `call_vsd_worker.py`   
- pass the `batch_renderer` to `pose_error.vsd(..)` instead of other renderers `ren`, and then compute vsd error in that function based on the renderer argument type.

Reply:
In my latest commit, I updated with your reviews except for the last point regarding batch renderer. As we discussed last week, I initially attempted to do exactly what you suggested but unfortunately, none of the renderers in the toolkit allow multiprocessing. Therefore, I had to find an alternative solution by running `call_vsd_worker.py` n_workers times.